File: sdk/ruby/test/test_coercion.rb
# frozen_string_literal: true

require_relative "test_helper"

class TestCoercion < Minitest::Test
  def test_string_to_int
    s = AnyVali.int_.coerce("string->int")
    assert_equal 42, s.parse("42")
  end

  def test_string_to_int_trims_whitespace
    s = AnyVali.int_.coerce("string->int")
    assert_equal 42, s.parse("  42  ")
  end

  def test_string_to_number
    s = AnyVali.number.coerce("string->number")
    assert_equal 3.14, s.parse("3.14")
  end

  def test_string_to_bool_true
    s = AnyVali.bool.coerce("string->bool")
    assert_equal true, s.parse("true")
  end

  def test_string_to_bool_false
    s = AnyVali.bool.coerce("string->bool")
    assert_equal false, s.parse("false")
  end

  def test_string_to_bool_1
    s = AnyVali.bool.coerce("string->bool")
    assert_equal true, s.parse("1")
  end

  def test_string_to_bool_0
    s = AnyVali.bool.coerce("string->bool")
    assert_equal false, s.parse("0")
  end

  def test_string_to_bool_case_insensitive
    s = AnyVali.bool.coerce("string->bool")
    assert_equal true, s.parse("TRUE")
  end

  def test_trim
    s = AnyVali.string.coerce("trim")
    assert_equal "hello", s.parse("  hello  ")
  end

  def test_lower
    s = AnyVali.string.coerce("lower")
    assert_equal "hello world", s.parse("HELLO World")
  end

  def test_upper
    s = AnyVali.string.coerce("upper")
    assert_equal "HELLO WORLD", s.parse("hello world")
  end

  def test_coercion_failure
    s = AnyVali.int_.coerce("string->int")
    result = s.safe_parse("not-a-number")
    assert result.failure?
    assert_equal "coercion_failed", result.issues.first.code
    assert_equal "int", result.issues.first.expected
    assert_equal "not-a-number", result.issues.first.received
  end

  def test_coercion_then_validation_fails
    s = AnyVali.int_.min(10).coerce("string->int")
    result = s.safe_parse("5")
    assert result.failure?
    assert_equal "too_small", result.issues.first.code
    assert_equal "10", result.issues.first.expected
    assert_equal "5", result.issues.first.received
  end

  def test_coercion_then_validation_success
    s = AnyVali.int_.min(1).max(100).coerce("string->int")
    assert_equal 50, s.parse("50")
  end

  def test_chained_coercions
    s = AnyVali.string.coerce(["trim", "lower"])
    assert_equal "hello", s.parse("  HELLO  ")
  end

  def test_int_passthrough_no_coerce
    s = AnyVali.int_.coerce("string->int")
    assert_equal 42, s.parse(42)
  end

  def test_coercion_config_portable
    assert AnyVali::CoercionConfig.portable?("trim")
    assert AnyVali::CoercionConfig.portable?("string->int")
    assert AnyVali::CoercionConfig.portable?(["trim", "lower"])
  end

  # --- Ergonomic no-arg / generic-"string" coercion -------------------------
  # The only portable coercion source is "string" (spec 5.1). Enabling coercion
  # on a numeric/bool schema with no explicit typed target (`.coerce` no-arg, or
  # the generic "string" source) infers the target from the schema kind. A bare
  # `.coerce` defaults the source to "string", so `number.coerce` and
  # `number.coerce("string")` are equivalent.

  def test_no_arg_coerce_equivalent_to_string
    no_arg = AnyVali.number.coerce
    explicit = AnyVali.number.coerce("string")
    assert_equal "string", no_arg.coerce_config
    assert_equal "string", explicit.coerce_config
    assert_equal 3.14, no_arg.parse("3.14")
  end

  def test_default_coerce_number_from_string
    s = AnyVali.number.coerce
    assert_equal 3.14, s.parse("3.14")
  end

  def test_default_coerce_int_from_string
    s = AnyVali.int_.coerce
    assert_equal 42, s.parse("42")
  end

  def test_default_coerce_bool_true_from_string
    s = AnyVali.bool.coerce
    assert_equal true, s.parse("true")
  end

  def test_default_coerce_bool_false_from_string
    s = AnyVali.bool.coerce
    assert_equal false, s.parse("false")
  end

  def test_default_coerce_object_numeric_fields_from_string
    s = AnyVali.object(
      properties: {
        "lumpSum" => AnyVali.number.coerce,
        "monthlyContributions" => AnyVali.number.coerce,
        "investmentTerm" => AnyVali.number.coerce
      },
      required: %w[lumpSum monthlyContributions investmentTerm]
    )

    result = s.safe_parse(
      "lumpSum" => "1000000",
      "monthlyContributions" => "1000",
      "investmentTerm" => "20"
    )

    assert result.success?, "object with default-coerce numeric fields must parse"
    assert_equal 1_000_000.0, result.value["lumpSum"]
    assert_equal 1_000.0, result.value["monthlyContributions"]
    assert_equal 20.0, result.value["investmentTerm"]
  end

  def test_no_arg_coerce_is_portable
    assert AnyVali::CoercionConfig.portable?("string")
    assert AnyVali.number.coerce.portable?
  end

  # --- Canonical coercion matrix (all FROM STRING, no-arg ergonomic) --------
  # Targets inferred from schema kind. Every ACCEPT/REJECT row from the spec.

  INT_ACCEPT = { "42" => 42, "  42  " => 42, "-7" => -7 }.freeze
  INT_REJECT = ["3.14", "0x10", "1_000", "+5", "Infinity", "", "abc"].freeze

  def test_matrix_string_to_int_accept
    s = AnyVali.int_.coerce
    INT_ACCEPT.each do |input, expected|
      result = s.safe_parse(input)
      assert result.success?, "int coerce must accept #{input.inspect}"
      assert_equal expected, result.value
    end
  end

  def test_matrix_string_to_int_reject
    s = AnyVali.int_.coerce
    INT_REJECT.each do |input|
      result = s.safe_parse(input)
      assert result.failure?, "int coerce must reject #{input.inspect}"
      assert_equal "coercion_failed", result.issues.first.code
    end
  end

  NUMBER_ACCEPT = { "3.14" => 3.14, "-1.5e3" => -1500.0, "  2  " => 2.0, "0" => 0.0 }.freeze
  NUMBER_REJECT = ["0x10", "Infinity", "NaN", "", "1_000", "abc"].freeze

  def test_matrix_string_to_number_accept
    s = AnyVali.number.coerce
    NUMBER_ACCEPT.each do |input, expected|
      result = s.safe_parse(input)
      assert result.success?, "number coerce must accept #{input.inspect}"
      assert_equal expected, result.value
    end
  end

  def test_matrix_string_to_number_reject
    s = AnyVali.number.coerce
    NUMBER_REJECT.each do |input|
      result = s.safe_parse(input)
      assert result.failure?, "number coerce must reject #{input.inspect}"
      assert_equal "coercion_failed", result.issues.first.code
    end
  end

  BOOL_TRUE = %w[true TRUE 1].freeze
  BOOL_FALSE = %w[false 0].freeze
  BOOL_REJECT = ["yes", "no", "on", "off", "t", "f", "2", ""].freeze

  def test_matrix_string_to_bool_true
    s = AnyVali.bool.coerce
    BOOL_TRUE.each do |input|
      result = s.safe_parse(input)
      assert result.success?, "bool coerce must accept #{input.inspect}"
      assert_equal true, result.value
    end
  end

  def test_matrix_string_to_bool_false
    s = AnyVali.bool.coerce
    BOOL_FALSE.each do |input|
      result = s.safe_parse(input)
      assert result.success?, "bool coerce must accept #{input.inspect}"
      assert_equal false, result.value
    end
  end

  def test_matrix_string_to_bool_reject
    s = AnyVali.bool.coerce
    BOOL_REJECT.each do |input|
      result = s.safe_parse(input)
      assert result.failure?, "bool coerce must reject #{input.inspect}"
      assert_equal "coercion_failed", result.issues.first.code
    end
  end

  # --- String-kind transforms (string source is a no-op; transforms apply) ---

  def test_string_kind_no_arg_coerce_is_noop_passthrough
    s = AnyVali.string.coerce
    assert_equal "hello", s.parse("hello")
  end

  def test_string_transforms_trim
    s = AnyVali.string.coerce("trim")
    assert_equal "hello", s.parse("  hello  ")
  end

  def test_string_transforms_chainable
    s = AnyVali.string.coerce(%w[trim lower upper])
    assert_equal "HELLO", s.parse("  Hello  ")
  end

  # CWE-20 / spec 5.1: non-portable coercion bypass. Ruby's Float() accepts
  # digit-group underscores ("1_000.5") and hex floats, which diverge from the
  # JS reference. string->number must accept ASCII decimals only.
  def test_string_to_number_rejects_non_decimal
    s = AnyVali.number.coerce("string->number")
    ["1_000.5", "0x1.8p3", "0x10"].each do |bad|
      assert s.safe_parse(bad).failure?, "string->number must reject #{bad.inspect}"
    end
    [["3.14", 3.14], ["+5", 5.0], [".5", 0.5], ["1e3", 1000.0]].each do |good, expected|
      result = s.safe_parse(good)
      assert result.success?, "string->number must accept #{good.inspect}"
      assert_equal expected, result.value
    end
  end
end

File: sdk/cpp/tests/test_coercion.cpp
#include "test_framework.hpp"
#include <anyvali/anyvali.hpp>

using namespace anyvali;
using json = nlohmann::json;

TEST("coercion: string->int valid") {
    auto s = int_();
    s->coerce("string->int");
    auto result = s->safe_parse(json("42"));
    ASSERT(result.success);
    ASSERT_JSON_EQ(result.value, json(42));
}

TEST("coercion: string->int trims whitespace") {
    auto s = int_();
    s->coerce("string->int");
    auto result = s->safe_parse(json("  42  "));
    ASSERT(result.success);
    ASSERT_JSON_EQ(result.value, json(42));
}

TEST("coercion: string->number valid") {
    auto s = number();
    s->coerce("string->number");
    auto result = s->safe_parse(json("3.14"));
    ASSERT(result.success);
    ASSERT_JSON_EQ(result.value, json(3.14));
}

TEST("coercion: string->bool true") {
    auto s = bool_();
    s->coerce("string->bool");
    ASSERT(s->safe_parse(json("true")).success);
    ASSERT_JSON_EQ(s->safe_parse(json("true")).value, json(true));
}

TEST("coercion: string->bool false") {
    auto s = bool_();
    s->coerce("string->bool");
    ASSERT_JSON_EQ(s->safe_parse(json("false")).value, json(false));
}

TEST("coercion: string->bool 1") {
    auto s = bool_();
    s->coerce("string->bool");
    ASSERT_JSON_EQ(s->safe_parse(json("1")).value, json(true));
}

TEST("coercion: string->bool 0") {
    auto s = bool_();
    s->coerce("string->bool");
    ASSERT_JSON_EQ(s->safe_parse(json("0")).value, json(false));
}

TEST("coercion: string->bool case insensitive") {
    auto s = bool_();
    s->coerce("string->bool");
    ASSERT_JSON_EQ(s->safe_parse(json("TRUE")).value, json(true));
}

TEST("coercion: trim removes whitespace") {
    auto s = string_();
    s->coerce("trim");
    auto result = s->safe_parse(json("  hello  "));
    ASSERT(result.success);
    ASSERT_JSON_EQ(result.value, json("hello"));
}

TEST("coercion: lower converts to lowercase") {
    auto s = string_();
    s->coerce("lower");
    auto result = s->safe_parse(json("HELLO World"));
    ASSERT(result.success);
    ASSERT_JSON_EQ(result.value, json("hello world"));
}

TEST("coercion: upper converts to uppercase") {
    auto s = string_();
    s->coerce("upper");
    auto result = s->safe_parse(json("hello world"));
    ASSERT(result.success);
    ASSERT_JSON_EQ(result.value, json("HELLO WORLD"));
}

TEST("coercion: failure produces coercion_failed") {
    auto s = int_();
    s->coerce("string->int");
    auto result = s->safe_parse(json("not-a-number"));
    ASSERT(!result.success);
    ASSERT(result.issues[0].code == "coercion_failed");
}

TEST("coercion: happens before validation") {
    auto s = int_();
    s->coerce("string->int");
    s->min(10);
    auto result = s->safe_parse(json("5"));
    ASSERT(!result.success);
    ASSERT(result.issues[0].code == "too_small");
}

TEST("coercion: then validation success") {
    auto s = int_();
    s->coerce("string->int");
    s->min(1);
    s->max(100);
    auto result = s->safe_parse(json("50"));
    ASSERT(result.success);
    ASSERT_JSON_EQ(result.value, json(50));
}

TEST("coercion: chained coercions left to right") {
    auto s = string_();
    s->coerce(std::vector<std::string>{"trim", "lower"});
    auto result = s->safe_parse(json("  HELLO  "));
    ASSERT(result.success);
    ASSERT_JSON_EQ(result.value, json("hello"));
}

// ---------------------------------------------------------------------------
// CWE-20 / spec 5.1: non-portable coercion bypass
//
// std::stoll/std::stod are more permissive than the ECMA-262 reference (JS):
// they accept a leading "+", and std::stod accepts hex floats ("0x1p4") and
// "inf"/"nan". Each let a string the JS reference rejects coerce into a number.
// Coercion must accept ASCII decimals only so behaviour matches across SDKs.
// ---------------------------------------------------------------------------

TEST("security: string->int rejects non-decimal forms") {
    auto s = int_();
    s->coerce("string->int");
    for (const auto& bad : {"+5", "1.0", "1e3", "0x10"}) {
        ASSERT(!s->safe_parse(json(bad)).success);
    }
    ASSERT(s->safe_parse(json("42")).success);
    ASSERT(s->safe_parse(json("  42  ")).success);
}

TEST("security: string->number rejects hex-float and infinities") {
    auto s = number();
    s->coerce("string->number");
    for (const auto& bad : {"0x1p4", "0x10", "inf", "nan", "infinity"}) {
        ASSERT(!s->safe_parse(json(bad)).success);
    }
    ASSERT(s->safe_parse(json("3.14")).success);
    ASSERT(s->safe_parse(json("+5")).success);
    ASSERT(s->safe_parse(json(".5")).success);
}

// ---------------------------------------------------------------------------
// Default / no-explicit-source coercion (cross-SDK regression).
//
// Background: the JS SDK lets you enable coercion via an object config whose
// source ("from") is optional; when omitted on a numeric/bool schema the
// coercion silently no-ops, so a string like "3.14" fails with invalid_type
// instead of being coerced. The only portable coercion source is "string".
//
// The C++ builder has NO separate no-source form: the *simplest* way to enable
// coercion is to name the portable token (e.g. coerce("string->int")), which
// intrinsically binds the source to "string". These tests exercise that
// default/idiomatic form end-to-end through safe_parse, mirroring the regression
// cases added in the other SDKs. They MUST coerce (not invalid_type).
// ---------------------------------------------------------------------------

TEST("default-coerce: number coerces \"3.14\" -> 3.14") {
    auto s = number();
    s->coerce("string->number");
    auto result = s->safe_parse(json("3.14"));
    ASSERT(result.success);
    ASSERT_JSON_EQ(result.value, json(3.14));
}

TEST("default-coerce: int coerces \"42\" -> 42") {
    auto s = int_();
    s->coerce("string->int");
    auto result = s->safe_parse(json("42"));
    ASSERT(result.success);
    ASSERT_JSON_EQ(result.value, json(42));
}

TEST("default-coerce: bool coerces \"true\"/\"false\"") {
    auto s = bool_();
    s->coerce("string->bool");
    auto t = s->safe_parse(json("true"));
    ASSERT(t.success);
    ASSERT_JSON_EQ(t.value, json(true));
    auto f = s->safe_parse(json("false"));
    ASSERT(f.success);
    ASSERT_JSON_EQ(f.value, json(false));
}

TEST("default-coerce: object with numeric fields coerces all string inputs") {
    auto s = object();
    s->prop("age", [] { auto a = int_(); a->coerce("string->int"); return a; }());
    s->prop("score", [] { auto sc = number(); sc->coerce("string->number"); return sc; }());
    s->prop("active", [] { auto b = bool_(); b->coerce("string->bool"); return b; }());
    s->required({"age", "score", "active"});

    json input = {{"age", "42"}, {"score", "3.14"}, {"active", "true"}};
    auto result = s->safe_parse(input);
    ASSERT(result.success);
    ASSERT_JSON_EQ(result.value["age"], json(42));
    ASSERT_JSON_EQ(result.value["score"], json(3.14));
    ASSERT_JSON_EQ(result.value["active"], json(true));
}

// ===========================================================================
// No-arg coerce() ergonomic + full FROM-STRING coercion matrix.
//
// coerce() with NO arguments enables string coercion with the target inferred
// from the schema kind:
//   - int-family kinds  -> "string->int"   (ASCII ^-?\d+$, trimmed)
//   - float/number kinds -> "string->number" (ASCII decimal float incl exponent)
//   - bool              -> "string->bool"   (trim + case-insensitive)
// String transforms (trim/lower/upper) stay explicit (string kind only) since
// they cannot be inferred from kind.
//
// Each ACCEPT/REJECT row of the canonical matrix is exercised below through the
// no-arg form end-to-end via safe_parse. Failures surface as coercion_failed.
// ===========================================================================

// ---- string -> int : ACCEPT "42","  42  ","-7" ----
TEST("matrix int noarg: accepts 42") {
    auto s = int_(); s->coerce();
    auto r = s->safe_parse(json("42"));
    ASSERT(r.success);
    ASSERT_JSON_EQ(r.value, json(42));
}

TEST("matrix int noarg: accepts padded 42") {
    auto s = int_(); s->coerce();
    auto r = s->safe_parse(json("  42  "));
    ASSERT(r.success);
    ASSERT_JSON_EQ(r.value, json(42));
}

TEST("matrix int noarg: accepts -7") {
    auto s = int_(); s->coerce();
    auto r = s->safe_parse(json("-7"));
    ASSERT(r.success);
    ASSERT_JSON_EQ(r.value, json(-7));
}

// ---- string -> int : REJECT "3.14","0x10","1_000","+5","Infinity","","abc" ----
TEST("matrix int noarg: rejects non-decimal forms") {
    auto s = int_(); s->coerce();
    for (const auto& bad : {"3.14", "0x10", "1_000", "+5", "Infinity", "", "abc"}) {
        auto r = s->safe_parse(json(bad));
        ASSERT(!r.success);
        ASSERT(r.issues[0].code == "coercion_failed");
    }
}

// ---- string -> number : ACCEPT "3.14","-1.5e3","  2  ","0" ----
TEST("matrix number noarg: accepts 3.14") {
    auto s = number(); s->coerce();
    auto r = s->safe_parse(json("3.14"));
    ASSERT(r.success);
    ASSERT_JSON_EQ(r.value, json(3.14));
}

TEST("matrix number noarg: accepts -1.5e3") {
    auto s = number(); s->coerce();
    auto r = s->safe_parse(json("-1.5e3"));
    ASSERT(r.success);
    ASSERT_JSON_EQ(r.value, json(-1500.0));
}

TEST("matrix number noarg: accepts padded 2") {
    auto s = number(); s->coerce();
    auto r = s->safe_parse(json("  2  "));
    ASSERT(r.success);
    ASSERT_JSON_EQ(r.value, json(2.0));
}

TEST("matrix number noarg: accepts 0") {
    auto s = number(); s->coerce();
    auto r = s->safe_parse(json("0"));
    ASSERT(r.success);
    ASSERT_JSON_EQ(r.value, json(0.0));
}

// ---- string -> number : REJECT "0x10","Infinity","NaN","","1_000","abc" ----
TEST("matrix number noarg: rejects non-decimal forms") {
    auto s = number(); s->coerce();
    for (const auto& bad : {"0x10", "Infinity", "NaN", "", "1_000", "abc"}) {
        auto r = s->safe_parse(json(bad));
        ASSERT(!r.success);
        ASSERT(r.issues[0].code == "coercion_failed");
    }
}

// ---- string -> bool : ACCEPT true<-"true","TRUE","1"; false<-"false","0" ----
TEST("matrix bool noarg: accepts true forms") {
    auto s = bool_(); s->coerce();
    for (const auto& good : {"true", "TRUE", "1"}) {
        auto r = s->safe_parse(json(good));
        ASSERT(r.success);
        ASSERT_JSON_EQ(r.value, json(true));
    }
}

TEST("matrix bool noarg: accepts false forms") {
    auto s = bool_(); s->coerce();
    for (const auto& good : {"false", "0"}) {
        auto r = s->safe_parse(json(good));
        ASSERT(r.success);
        ASSERT_JSON_EQ(r.value, json(false));
    }
}

TEST("matrix bool noarg: trims and is case-insensitive") {
    auto s = bool_(); s->coerce();
    ASSERT_JSON_EQ(s->safe_parse(json("  TrUe ")).value, json(true));
    ASSERT_JSON_EQ(s->safe_parse(json(" FALSE ")).value, json(false));
}

// ---- string -> bool : REJECT "yes","no","on","off","t","f","2","" ----
TEST("matrix bool noarg: rejects ambiguous forms") {
    auto s = bool_(); s->coerce();
    for (const auto& bad : {"yes", "no", "on", "off", "t", "f", "2", ""}) {
        auto r = s->safe_parse(json(bad));
        ASSERT(!r.success);
        ASSERT(r.issues[0].code == "coercion_failed");
    }
}

// ---- no-arg target inference per kind ----
TEST("matrix noarg: int-family infers string->int (rejects float text)") {
    // int8/int32/uint16 etc. must reject "3.14" like plain int does.
    {
        auto s = int32(); s->coerce();
        ASSERT(!s->safe_parse(json("3.14")).success);
        ASSERT_JSON_EQ(s->safe_parse(json("5")).value, json(5));
    }
    {
        auto s = uint16(); s->coerce();
        ASSERT(!s->safe_parse(json("3.14")).success);
        ASSERT_JSON_EQ(s->safe_parse(json("5")).value, json(5));
    }
}

TEST("matrix noarg: float32 infers string->number (accepts float text)") {
    auto s = float32(); s->coerce();
    auto r = s->safe_parse(json("3.14"));
    ASSERT(r.success);  // float kind -> string->number accepts decimals
}

// ---- no-arg form composes with downstream numeric validation ----
TEST("matrix noarg: coercion happens before range validation") {
    auto s = int_(); s->coerce(); s->min(10);
    auto r = s->safe_parse(json("5"));
    ASSERT(!r.success);
    ASSERT(r.issues[0].code == "too_small");  // coerced to 5, then range fails
}

TEST("matrix noarg: typed-token overload still works alongside") {
    auto s = int_(); s->coerce("string->int");
    ASSERT_JSON_EQ(s->safe_parse(json("42")).value, json(42));
}

// ---- string transforms stay explicit (string kind only), chainable ----
TEST("matrix string transforms: trim/lower/upper chainable") {
    {
        auto s = string_(); s->coerce("trim");
        ASSERT_JSON_EQ(s->safe_parse(json("  hi  ")).value, json("hi"));
    }
    {
        auto s = string_(); s->coerce("lower");
        ASSERT_JSON_EQ(s->safe_parse(json("HeLLo")).value, json("hello"));
    }
    {
        auto s = string_(); s->coerce("upper");
        ASSERT_JSON_EQ(s->safe_parse(json("HeLLo")).value, json("HELLO"));
    }
    {
        auto s = string_(); s->coerce(std::vector<std::string>{"trim", "upper"});
        ASSERT_JSON_EQ(s->safe_parse(json("  hi  ")).value, json("HI"));
    }
}
